Refactor info plugin codebase

pmuens opened the pull request:
## What did you implement:

Refactoring of the `info` plugin codebase.

## How did you implement it:

Refactored the `info` plugin codebase.

## How can we verify it:

Just run `serverless info`

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

***Is this ready for review?:*** YES

eahefnawy commented on lib/plugins/aws/info/index.js:
```diff
@@ -184,40 +198,203 @@ ${chalk.yellow('region:')} ${info.region}`;
+            if (metric.Label === 'Invocations') {
+              functionsMessage += `      ${chalk.yellow('invocations:')}`;
+              if (metric.Datapoints.length) {
+                functionsMessage += ` ${metric.Datapoints[0].Sum}x`;
```
imo let's ditch the `x` ... so it'd be `invocations: 2`

Also, is there a reason why it's scoped only to the last 24h?!

pmuens replied:
Yes, we need to make 4 API calls for each function. Furthermore the more data we get back, the slower it gets.

I'll add another `metrics` plugin where one can investigate further, but the `info` command should display some more general info IMHO.

mthenw replied:
> imo let's ditch the x ... so it'd be invocations: 2

+1

mthenw commented on lib/plugins/aws/info/index.js:
```diff
@@ -184,40 +198,203 @@ ${chalk.yellow('region:')} ${info.region}`;
+              if (metric.Datapoints.length) {
+                functionsMessage += ` ${metric.Datapoints[0].Sum}`;
+              } else {
+                functionsMessage += ' None';
```
why "None" instead of 0?

mthenw commented on lib/plugins/aws/info/index.js:
```diff
@@ -184,40 +198,203 @@ ${chalk.yellow('region:')} ${info.region}`;
+              if (metric.Datapoints.length) {
+                const avgDuration = metric.Datapoints[0].Average;
+                const roundedAvgDuration = Math.round(avgDuration * 100) / 100;
+                functionsMessage += ` ${roundedAvgDuration}ms avg.`;
```
> duration: 2.44ms avg.

IMHO

> avg duration: 2.44ms

pmuens commented:
@eahefnawy @mthenw thanks for your comments!

I agree and have fixed everything. Furthermore I added 100% test coverage. I had to split the whole `awsInfo` plugin up into separate files, because otherwise testing this thing was impossible.
I rewrote each and every test from scratch so that we have 100% test coverage of everything and will be on the safe side for future enhancements.

This will make the review process a little bit harder but will make everything a lot easier in the future as it will prevent conflicts.

The best way to test this now is to deploy a service a play around with it a little bit.

pmuens commented:
Will update this PR so that the metrics implementation is removed and we only got the refactoring in place.

pmuens commented on lib/plugins/aws/info/display.js:
```diff
@@ -0,0 +1,83 @@
+    message += endpointsMessage;
+
+    // Display function information
+    let functionsMessage = `\n${chalk.yellow('functions:')}`;
```
Add a `\n` at the end here and...

pmuens commented on lib/plugins/aws/info/display.js:
```diff
@@ -0,0 +1,83 @@
+
+    if (info.functions && info.functions.length > 0) {
+      info.functions.forEach((f) => {
+        functionsMessage += `\n  ${f.name}:\n`;
```
... remove the `\n` here so that functions are displayed close to each other like:

```
functions:
  func1:
    arn:foo:bar
  func2:
    arn:foo:bar
```

pmuens commented:
@nikgraf @eahefnawy @mthenw ready for a final review. This is just a refactoring of the codebase. I've removed the metrics part. Should be merged ASAP to prevent future conflicts...

nikgraf commented:
Reviewing this right now …

pmuens commented on lib/plugins/aws/info/display.js:
```diff
@@ -60,11 +60,10 @@ module.exports = {
       info.functions.forEach((f) => {
-        functionsMessage += `\n  ${f.name}:\n`;
-        functionsMessage += `    ${f.arn}`;
+        functionsMessage = functionsMessage.concat(`\n  ${f.name}: ${f.arn}`);
```
Just one quick question. Why using concat?

We should IMHO use one way of concatenating and use it throughout the whole info codebase.

pmuens commented on lib/plugins/aws/info/display.js:
```diff
@@ -0,0 +1,82 @@
+
+    if (info.functions && info.functions.length > 0) {
+      info.functions.forEach((f) => {
+        functionsMessage = functionsMessage.concat(`\n  ${f.name}: ${f.arn}`);
```
IMHO we should use one way to concatenate strings.

`+=` or `concat`. But stick to it throughout the whole `info` codebase.

nikgraf replied:
done

nikgraf commented:
Good to be merged after the tests succeed

pmuens commented:
Thanks for the review @nikgraf 

merging...
